src/vm/contexts.rs

```diff
@@ -404,6 +403,22 @@ impl <'a,'b> Environment <'a,'b> {
         }
     }
 
+    pub fn evaluate_at_block(&mut self, bhh: BlockHeaderHash, closure: &SymbolicExpression, local: &LocalContext) -> Result<Value> {
+        self.global_context.begin_read_only();
+
+        let result = self.global_context.database.set_block_hash(bhh)
+            .and_then(|prior_bhh| {
+                let result = eval(closure, self, local);
+                self.global_context.database.set_block_hash(prior_bhh)
+                    .expect("ERROR: Failed to restore prior active block after time-shifted evaluation.");
+                result
+            });
+
+        self.global_context.roll_back();
```

Comment:
It's just an additional safety mechanism for enforcing the read-only nature of that execution. Shouldn't ever be necessary (because there are both the analysis checks and other runtime checks in place to enforce read-only contexts).